src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -183,7 +183,8 @@ void IDeserializationCallback.OnDeserialization(object sender)
 		internal SessionImpl(SessionFactoryImpl factory, ISessionCreationOptions options)
 			: base(factory, options)
 		{
-			using (BeginContext())
+			_context = SessionIdLoggingContext.CreateOrNull(SessionId);
```

Comment:
I think we should explain here in code why it is not disposed immediately. Something like:
```c#
// This context is disposed only on session own disposal. This greatly reduces the number of context switches
// for most usual session usages. It may cause an irrelevant session id to be set back on disposal, but since all
// session entry points are supposed to set it, it should not have any consequences.
```

Reply:
Done